src/Admin/Controller/RoleController.php

```diff
@@ -48,20 +56,54 @@ public function delete(Role $role): Response
 
     private function handleForm(Request $request, ?Role $role = null): Response
     {
-        $form = $this->createForm(RoleType::class, $role);
+        $permissions = $this->permissionRepository->findAll();
+        $form = $this->createForm(RoleType::class, $role, ['permissions' => $permissions]);
 
         $form->handleRequest($request);
         if ($form->isSubmitted() && $form->isValid()) {
             $role = $form->getData();
+
+            foreach ($permissions as $permission) {
+                $permissionId = 'permission_' . $permission->getId();
+                $hasPermission = $form->get($permissionId)->getData();
+
+                if ($hasPermission && !$role->getPermissions()->contains($permission)) {
+                    $role->addPermission($permission);
+                } elseif (!$hasPermission && $role->getPermissions()->contains($permission)) {
+                    $role->removePermission($permission);
+                }
+            }
+
             $this->roleRepository->save($role);
 
             $this->addFlash('success', 'flashes.role_saved');
             return $this->redirectToRoute('forumify_admin_role_list');
         }
 
+        $plugins = $this->pluginRepository->findAll();
```

Comment:
This will have to be modified anyway if perms don't come from the database anymore, but move this logic to the `RoleType`. You can inject services like the `PluginService` into the `RoleType`, iterate over the ForumifyBundle + all active plugins, and add fields for them directly